edge/app.go

```diff
@@ -0,0 +1,106 @@
+package edge
+
+import (
+	"fmt"
+
+	"github.com/futurehomeno/fimpgo"
+	"github.com/futurehomeno/fimpgo/discovery"
+
+	"github.com/futurehomeno/cliffhanger/lifecycle"
+	"github.com/futurehomeno/cliffhanger/router"
+	"github.com/futurehomeno/cliffhanger/task"
+)
+
+// Edge is an interface representing an edge application service.
+type Edge interface {
+	// Start starts the edge application.
+	Start() error
+	// Stop stops the edge application maintaining a graceful shutdown.
+	Stop() error
+}
+
+// New creates a new edge application instance.
+func New(
+	mqtt *fimpgo.MqttTransport,
+	discovery *discovery.ServiceDiscoveryResponder,
+	lifecycle *lifecycle.Lifecycle,
+	topicSubscriptions []string,
+	router router.Router,
+	taskManager task.Manager,
+) Edge {
+	return &edge{
+		mqtt:               mqtt,
+		discovery:          discovery,
+		lifecycle:          lifecycle,
+		topicSubscriptions: topicSubscriptions,
+		messageRouter:      router,
+		taskManager:        taskManager,
+	}
+}
+
+// edge is an implementation of edge application interface.
+type edge struct {
```

Comment:
Nice try but no dice! This is Golang. You can embed and inherit methods, but you can't overload them. Both edge and core have methods Start and Stop, which have slightly different behaviors. The only way to do it, would be to create a third struct called "common", implement additional methods there and NOT EMBED but incorporate as private in the edge and create respectively.

This complication is not worth the hassle because most likely core and edge will diverge even further. So in this case I prefer a KISS over total DRY. But nice try!

Reply:
That was the idea I was thinking about: https://play.golang.org/p/siIo9ZSLR9Q
But if both `edge` and `core` are going to evolve independently from each other, then it's fine.